packages/contract-wrappers/src/contract_wrappers/forwarder_wrapper.ts

```diff
@@ -0,0 +1,209 @@
+     * 5% of ETH value is reserved for paying fees to order feeRecipients (in ZRX) and forwarding contract feeRecipient (in ETH).
+     * Any ETH not spent will be refunded to sender.
+     * @param   signedOrders         An array of objects that conform to the SignedOrder interface. All orders must specify the same makerAsset.
+     *                               All orders must specify WETH as the takerAsset
+     * @param   takerAddress         The user Ethereum address who would like to fill this order. Must be available via the supplied
+     *                               Provider provided at instantiation.
+     * @param   ethAmount            The amount of eth to send with the transaction (in wei).
+     * @param   signedFeeOrders      An array of objects that conform to the SignedOrder interface. All orders must specify ZRX as makerAsset and WETH as takerAsset.
+     *                               Used to purchase ZRX for primary order fees.
+     * @param   feePercentage        The percentage of WETH sold that will payed as fee to forwarding contract feeRecipient.
+     *                               Defaults to 0.
+     * @param   feeRecipientAddress  The address that will receive ETH when signedFeeOrders are filled.
+     * @param   txOpts               Transaction parameters.
+     * @return  Transaction hash.
+     */
+    public async marketSellOrdersWithEthAsync(
+        signedOrders: SignedOrder[],
+        takerAddress: string,
+        ethAmount: BigNumber,
+        signedFeeOrders: SignedOrder[] = [],
+        feePercentage: BigNumber = constants.ZERO_AMOUNT,
+        feeRecipientAddress: string = constants.NULL_ADDRESS,
+        txOpts: TransactionOpts = {},
+    ): Promise<string> {
+        // type assertions
+        assert.doesConformToSchema('signedOrders', signedOrders, schemas.signedOrdersSchema);
+        await assert.isSenderAddressAsync('takerAddress', takerAddress, this._web3Wrapper);
+        assert.isBigNumber('ethAmount', ethAmount);
+        assert.doesConformToSchema('signedFeeOrders', signedFeeOrders, schemas.signedOrdersSchema);
+        assert.isBigNumber('feePercentage', feePercentage);
+        assert.isETHAddressHex('feeRecipientAddress', feeRecipientAddress);
+        assert.doesConformToSchema('txOpts', txOpts, txOptsSchema);
+        // other assertions
+        assert.ordersCanBeUsedForForwarderContract(signedOrders, this.getEtherTokenAddress());
+        assert.feeOrdersCanBeUsedForForwarderContract(
+            signedFeeOrders,
+            this.getZRXTokenAddress(),
+            this.getEtherTokenAddress(),
+        );
+        const normalizedTakerAddress = takerAddress.toLowerCase();
+        const normalizedFeeRecipientAddress = feeRecipientAddress.toLowerCase();
+        const forwarderContractInstance = await this._getForwarderContractAsync();
+        const txHash = await forwarderContractInstance.marketSellOrdersWithEth.sendTransactionAsync(
+            signedOrders,
```

Comment:
We should optimize this calldata before sending the transaction. The forwarding contract makes a lot of assumptions about the `assetData` fields that allow us to pass in empty byte arrays and save on gas.

For `orders`, all of the `takerAssetData` fields and all but the first order's `makerAssetData` fields can be empty. For all of the `feeOrders`, all of the `makerAssetData` and `takerAssetData` fields can be empty.